third_party/proton/csrc/lib/Data/TraceData.cpp

```diff
@@ -236,68 +236,93 @@ void TraceData::clear() {
 
 namespace {
 
-struct CycleMetricInternal {
-  CycleMetricInternal(uint32_t start, uint32_t end, uint32_t block,
-                      uint32_t unit, uint32_t proc, int32_t context,
-                      uint32_t timeShift)
-      : startCycle(start), endCycle(end), blockId(block), unitId(unit),
-        procId(proc), contextId(context), timeShiftCost(timeShift) {}
-
-  uint32_t startCycle;
-  uint32_t endCycle;
-  uint32_t blockId;
-  uint32_t unitId;
-  uint32_t procId;
+// Structure to pair CycleMetric with its context for processing
+struct CycleMetricWithContext {
+  std::shared_ptr<CycleMetric> cycleMetric;
   uint32_t contextId;
-  uint32_t timeShiftCost;
+  
+  CycleMetricWithContext(std::shared_ptr<CycleMetric> metric, uint32_t ctx)
+      : cycleMetric(metric), contextId(ctx) {}
 };
 
 std::vector<KernelTrace>
 convertToTimelineTrace(TraceData::Trace *trace,
-                       const std::map<std::string, int> &kernelBlockNum,
-                       std::vector<CycleMetricInternal> &cycleEvents) {
+                       std::vector<CycleMetricWithContext> &cycleEvents) {
   std::vector<KernelTrace> results;
-  results.reserve(kernelBlockNum.size());
```

Comment:
I feel like reserving the space is a bit overkill and could complicate the logic. We do not care about the performance of the finalization phase usually.